File: backend/geoEvent/src/UserService/UserService.Domain/Entities/Report.cs
using UserService.Domain.Enums;

namespace UserService.Domain.Entities;

public class Report
{
    public int ReportId { get; private set; }
    public ReportTargetType TargetType { get; private set; }
    public int? TargetId { get; private set; }
    public string Reason { get; private set; } = string.Empty;
    public string? Description { get; private set; }
    public ReportStatus Status { get; private set; }
    public int ReporterId { get; private set; }
    public int? ResolvedById { get; private set; }
    public string? ResolutionNote { get; private set; }
    public string? ModeratorAction { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime? ResolvedAt { get; private set; }

    public User? Reporter { get; private set; }
    public User? ResolvedBy { get; private set; }

    public User? TargetUser { get; private set; }

    protected Report() { }

    public Report(
        ReportTargetType targetType,
        int targetId,
        string reason,
        int reporterId,
        string? description = null)
    {
        TargetType = targetType;
        TargetId = targetId;
        Reason = reason.Trim();
        ReporterId = reporterId;
        Description = string.IsNullOrWhiteSpace(description) ? null : description.Trim();
        Status = ReportStatus.Pending;
        CreatedAt = DateTime.UtcNow;
    }

    public void MarkUnderReview(int adminUserId, string? resolutionNote, string? moderatorAction)
    {
        Status = ReportStatus.UnderReview;
        ResolvedById = adminUserId;
        ResolvedAt = null;
        ResolutionNote = string.IsNullOrWhiteSpace(resolutionNote) ? null : resolutionNote.Trim();
        ModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void Resolve(int adminUserId, string? resolutionNote, string? moderatorAction)
    {
        Status = ReportStatus.Resolved;
        ResolvedById = adminUserId;
        ResolvedAt = DateTime.UtcNow;
        ResolutionNote = string.IsNullOrWhiteSpace(resolutionNote) ? null : resolutionNote.Trim();
        ModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }

    public void Dismiss(int adminUserId, string? resolutionNote, string? moderatorAction)
    {
        Status = ReportStatus.Dismissed;
        ResolvedById = adminUserId;
        ResolvedAt = DateTime.UtcNow;
        ResolutionNote = string.IsNullOrWhiteSpace(resolutionNote) ? null : resolutionNote.Trim();
        ModeratorAction = string.IsNullOrWhiteSpace(moderatorAction) ? null : moderatorAction.Trim();
    }
}

File: frontend/geo_event/lib/shared/reports/models/report.dart
import 'report_status.dart';
import 'report_target_type.dart';

class Report {
  final int reportId;
  final ReportTargetType targetType;
  final int? targetId;
  final String reason;
  final String description;
  final ReportStatus status;
  final int? reporterId;
  final int? resolvedById;
  final DateTime createdAt;
  final DateTime? resolvedAt;

  const Report({
    required this.reportId,
    required this.targetType,
    required this.targetId,
    required this.reason,
    required this.description,
    required this.status,
    required this.reporterId,
    required this.resolvedById,
    required this.createdAt,
    required this.resolvedAt,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      reportId: (json['reportId'] as num).toInt(),
      targetType: ReportTargetType.fromJson(
        json['targetType'] as String? ?? 'Event',
      ),
      targetId: (json['targetId'] as num?)?.toInt(),
      reason: json['reason'] as String? ?? '',
      description: json['description'] as String? ?? '',
      status: ReportStatus.fromJson(json['status'] as String? ?? 'Pending'),
      reporterId: (json['reporterId'] as num?)?.toInt(),
      resolvedById: (json['resolvedById'] as num?)?.toInt(),
      createdAt: DateTime.parse(json['createdAt'] as String),
      resolvedAt: json['resolvedAt'] == null
          ? null
          : DateTime.parse(json['resolvedAt'] as String),
    );
  }
}